public/app/plugins/panel/table2/module.tsx

```diff
@@ -1,11 +1,40 @@
 import { PanelPlugin } from '@grafana/data';
-
-import { TablePanelEditor } from './TablePanelEditor';
 import { TablePanel } from './TablePanel';
-import { tableFieldRegistry } from './custom';
 import { Options, defaults } from './types';
 
 export const plugin = new PanelPlugin<Options>(TablePanel)
   .setDefaults(defaults)
-  .setCustomFieldConfigs(tableFieldRegistry)
-  .setEditor(TablePanelEditor);
+  .setCustomFieldConfigs(builder => {
+    builder
+      .addNumberInput({
+        id: 'width',
+        name: 'Column width',
+        description: 'column width (for table)',
+        settings: {
+          placeholder: 'auto',
+          min: 20,
+          max: 300,
+        },
+      })
```

Comment:
Amazing work on this as usual. I really like the fluent API, I think it looks great. Not sure I can give a fair review of the whole picture here, but I have the following question/thought.

I'm guessing the example in public/app/plugins/panel/table2/module.tsx is like that a plugin author would write. Would it make more sense for a plugin author to define the DTO/fields the author needs and have a default editor supplied by Grafana for each of the fields in that DTO.

Then the extension of that would be that if the plugin author is not happy with the default editors, the plugin author could easily add overrides in this fashion.

So instead of `. addNumberInput` we could supply `. addNumberField`. I realize now that my question seems very similar to what this api is, except for separating the fields from the visualization part but maybe that separation just adds complexity.

Not sure my question/rambling makes any sense at all and prob too late for this kind of input. Again, awesome work on this! 🎉

Reply:
I think this sort of makes sense, as a string value could have an "editor" component using a text input, select dropdown with options, or a radio button group. They would require different settings (at least select and radio group requires options).

So I sort of think this more UI centric approach makes sense but not sure, we have tried more "declarative data models" approaches before (dominik done 2 massive POCs on it and they all become a bit too complex)

Reply:
Yeah, with DTOs it's becoming massively more complex, and the benefits do not really resonate with what (I) think might be relevant for the API consumer. This approach though is way simpler compared to what I've tried to achieve with the previous POCs (https://github.com/grafana/grafana/pull/16773, https://github.com/grafana/grafana/pull/16731), i.e it does not play at all with nested options objects (you need to handle this with custom editor). It compromise some flexibility or data driven options, but I'm super happy to revisit that approach in the future.